src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/MarkupTransitionCompletionItemProvider.cs

```diff
@@ -0,0 +1,113 @@
+        public static RazorCompletionItem MarkupTransitionCompletionItem
+        {
+            get
+            {
+                if (_markupTransitionCompletionItem == null)
+                {
+                    var completionDisplayText = SyntaxConstants.TextTagName;
+                    _markupTransitionCompletionItem = new RazorCompletionItem(
+                        completionDisplayText,
+                        completionDisplayText,
+                        RazorCompletionItemKind.MarkupTransition,
+                        ElementCommitCharacters);
+                    var completionDescription = new MarkupTransitionCompletionDescription(Resources.MarkupTransition_Description());
+                    _markupTransitionCompletionItem.SetMarkupTransitionCompletionDescription(completionDescription);
+                }
+
+                return _markupTransitionCompletionItem;
+            }
+        }
+
+        public MarkupTransitionCompletionItemProvider(HtmlFactsService htmlFactsService)
+        {
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorSyntaxTree syntaxTree, TagHelperDocumentContext tagHelperDocumentContext, SourceSpan location)
+        {
+            if (syntaxTree is null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var change = new SourceChange(location, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!AtMarkupTransitionCompletionPoint(owner))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var parent = owner.Parent;
+            if (!_htmlFactsService.TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) ||
+                !containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var completions = new List<RazorCompletionItem>() { MarkupTransitionCompletionItem };
+            return completions;
+        }
+
+        // Internal for testing
+        internal static bool AtMarkupTransitionCompletionPoint(RazorSyntaxNode owner)
+        {
+            // Only provide IntelliSense for C# code blocks, of the form:
+            // @{ }, @code{ }, @functions{ }, @if(true){ }
+            var closestSignificantAncestor = owner.Ancestors().FirstOrDefault(node =>
+            {
+                if (node is MarkupElementSyntax markupNode && markupNode.ChildNodes().Count != 1)
```

Comment:
Does this work for the following example?
- Type `@{ <te`
- Press Escape key to close completion that pops up
- Now with the cursor after `e`, press `ctrl+space` to manually invoke completion

Does `<text>` completion show up?

Reply:
Also, `ChildNodes()` method doesn't give you the Children of the element, but it just gives you the different parts that make up a SyntaxNode. 
Example,
- For `<` --> ChildNodes().Count of the encompassing `MarkupElementSyntax` is 1
- For `<></> --> ChildNodes().Count` of the encompassing `MarkupElementSyntax` is 2
So, given your condition, if you invoke completion in the start tag above, it will incorrectly not give `<text>` completion

Reply:
As discussed, created a separate issue to track this:
https://github.com/dotnet/aspnetcore/issues/19815